Add support for smart playlists

chme opened the pull request:
These commits add support for smart playlists to forked-daapd. If a file is a smart playlist is identified by the filename ending with ".smartpl". The filescanner parses the file contents and creates the WHERE clause for the files table (this is similar to the existing system special playlists). The parsing is done with antlr3 (grammar files are "SMARTPL.g" and "SMARTPL2SQL.g"). 

The syntax is really close to the mt-daapd smart playlist syntax (
http://sourceforge.net/p/mt-daapd/code/HEAD/tree/tags/release-0.2.4.2/contrib/mt-daapd.playlist).

Differences are:
* only one smart playlist per file
* the not operator must be placed before an expression and not before the operator
* "||", "&&", "!" are not supported (use "or", "and", "not")
* comments are not supported

Following tags (with their types) are supported:
* artist (string)
* album_artist (string)
* album (string)
* title (string)
* genre (string)
* composer (string)
* path (string)
* type (string)
* data_kind (integer)
* media_kind (integer)
* play_count (integer)
* rating (integer)
* year (integer)
* compilation (integer)
* time_added (date)
* time_played (date)

Examples:
"recently added" { time_added after last week }

"60's Music" { year >= 1960 and year < 1970 and media_kind = 1 }

"rock music" { genre includes "rock" and media_kind = 1 }

chme commented:
@ejurgensen it would be great, if you could give this a try (even after writing the grammar files, i still do not fully understand how antlr3 works) and have a look at the changes i made. Especially to the existing code:

* I changed the meaning of PL_SMART to be used for the newly introduced smart playlists and added a new playlist type PL_SPECIAL for the former library smart playlists. I could not think of a different name that would not lead to confusion. (first commit)
* The filelist view needed to be changed, so that the smart playlists show up over the mpd protocol. I added this to the db upgrade v17, instead of introducing a new db version. (third commit)

ejurgensen commented on src/httpd_daap.c:
```diff
@@ -1685,8 +1685,8 @@ daap_reply_playlists(struct evhttp_request *req, struct evbuffer *evbuf, char **
-      /* Don't add empty Smart playlists */
-      if ((plid > 1) && (plitems == 0) && (pltype == PL_SMART))
+      /* Don't add empty Special playlists */
+      if ((plid > 1) && (plitems == PL_SPECIAL) && (pltype == 1))
```
Looks like items and type got swapped here?

ejurgensen commented on src/db.h:
```diff
@@ -166,11 +166,12 @@ struct media_file_info {
   PL_FOLDER = 1,
-  PL_SMART = 2,
+  PL_SPECIAL = 2,
+  PL_SMART = 3,
```
Playlists will be sent to the client according to "f.type DESC", so this way it will be smart playlists before the special ones. I guess that will work, but I would prefer keeping the same order as iTunes, which I think begins with the special ones. My suggestion is to change the sort to "f.type ASC", and then have: special (0), folder (1) and then either smart or plain. Some clients (Retune) require folder pl's to come before their children pl's. However that does require a db upgrade. What do you think?

chme commented:
Changing the order of the playlist types makes sense. The last two commits change the order to
* PL_SPECIAL
* PL_FOLDER
* PL_SMART
* PL_PLAIN

I modified the db upgrade v17 to address the pl_type changes.

ejurgensen commented:
I think you should make a v18 and not modify the v17 db upgrade, because doing that will leave people (like me) who already upgraded to v17 stranded. In hind-sight I do regret the v17 upgrade, though. Should have kept the special pl's as 0 and changed the ordering back then. Too late now.

You also need to add pli->type = PL_PLAIN explicitly everywhere there is a db_pl_add(). Now that plain pl's no longer are zero, the memset won't work.

chme commented:
The pl_type change is now done in the v18 db upgrade. I hope, i found all places where pli->type = PL_PLAIN needs to be set. 

I changed the syntax for data_kind and media_kind, now the user does not need to know which int value represents e. g. podcasts. Additionally i introduced enums for data_kind and media_kind, i think this improves the readability of the code.

It would be great, if you could give these changes another review. 
I will do some more tests over the coming weekend.

ejurgensen commented:
This is really great, I like everything about it. Both the feature itself, which I think quite a few will enjoy, and also that you took the time improve the code quality with the enumerations. And also nice to (finally) close the gap towards mt-daapd.

I expect to make a new release (23.0) within the next week, and I will also publish this version for the RPi. It will have a lot of good stuff thanks to your work. It would be good with some thorough testing before that. I will of course also run some tests.
